### PR TITLE
Broken PR: RangeFog and DensityFog

### PR DESCRIPTION
**Pull request by EliasHasle**
**Update: #17592 replaces this PR.**

Step 2 of the plan in #17264, adapted to comply with https://github.com/mrdoob/three.js/issues/17420#issuecomment-534592079.

<details><summary>Outdated</summary>Also correct typescript docstring and misspelled name in test file for FogExp2, as well as docs for standard Fog.</details>

Includes the change to distance-based depth, as well as corrections to better support mediump and avoid unnecessary squaring of a square root. Makes #17316 and #17324 obsolete.

Requires new build, update to the library build used in the editor, doc build, examples build.

I added an example named `webgl_geometry_terrain_fog_types`, which is basically `webgl_geometry_terrain_fog` without flight and with a GUI to try different fog setups. Everything seems to work. <details><summary>Outdated link... Will be updated soon</summary>Running version here: https://eliashasle.github.io/testing/fog_types/ (I think `FogExp` looks much more natural than the others.)</details>

I have added additional versions of the new example, that test legacy signatures, `mediump` and `lowp`, respectively. They should probably be removed before merging.

I am currently struggling with git. Hoping to get everything in order soon.

<details><summary>Resolved</summary>Some examples give an apparently unrelated warning message about using `pow` with negative input. I would really appreciate pointers into what goes wrong, as I am not using `pow` in my code or, as far as I can see, outputting anything that can be negative.</details>

### CODE REVIEW

**Comment by aardgoose** [comment]
See #4834 for discussion about the pow() warnings, it's a known issue, and actually a DirectX warning from the Angle translation of shaders.

**Comment by EliasHasle** [comment]
@aardgoose Good to know it is a known problem, at least. Not introduced by me now.

This is (properly) ready for review.

**Comment by EliasHasle** [comment]
@gogoend This PR modifies existing docs and adds a new class `FogExp`, but adds no corresponding Chinese doc page. I linked to the English version in the list, as a fallback.

File: docs/api/en/scenes/Fog.html [hunk 1]
```diff
@@ -10,7 +10,7 @@
 	<body>
 		<h1>[name]</h1>
 
-		<p class="desc">This class contains the parameters that define linear fog, i.e., that grows linearly denser with the distance.</p>
+		<p class="desc">This class contains the parameters that define standard fog, i.e., that grows smoothly denser from the near to the far distance.</p>
```
Comment (EliasHasle) [inline docs/api/en/scenes/Fog.html]: I am not sure about calling this "standard", but I really panicked when I saw that it wasn't linear at all. The smoothstep function is a cubic Hermite spline with clamped input.

File: docs/list.js [hunk 2]
```diff
@@ -764,6 +765,7 @@ var list = {
 
 			"场景": {
 				"Fog": "api/zh/scenes/Fog",
+				"FogExp": "api/en/scenes/FogExp", //ZH doc missing for now
```
Comment (EliasHasle) [inline docs/list.js]: I am not sure temporarily linking to English docs is an acceptable solution.

File: editor/js/Sidebar.Scene.js [hunk 3]
```diff
@@ -143,8 +143,9 @@ Sidebar.Scene = function ( editor ) {
 	var fogType = new UI.Select().setOptions( {
 
 		'None': 'None',
-		'Fog': 'Linear',
-		'FogExp2': 'Exponential'
+		'Fog': 'Ranged',
```
Comment (EliasHasle) [inline editor/js/Sidebar.Scene.js]: This time I renamed "Linear" to "Ranged". Maybe that is better. But it is inconsistent.

File: editor/js/Sidebar.Scene.js [hunk 4]
```diff
@@ -261,7 +267,7 @@ Sidebar.Scene = function ( editor ) {
 		fogPropertiesRow.setDisplay( type === 'None' ? 'none' : '' );
 		fogNear.setDisplay( type === 'Fog' ? '' : 'none' );
 		fogFar.setDisplay( type === 'Fog' ? '' : 'none' );
-		fogDensity.setDisplay( type === 'FogExp2' ? '' : 'none' );
+		fogDensity.setDisplay( (type === 'FogExp' || type === 'FogExp2') ? '' : 'none' );
```
Comment (EliasHasle) [inline editor/js/Sidebar.Scene.js]: Formatting mistake here. (Will fix soon.)

File: examples/files.js [hunk 5]
```diff
@@ -42,6 +42,7 @@ var files = {
 		"webgl_geometry_teapot",
 		"webgl_geometry_terrain",
 		"webgl_geometry_terrain_fog",
+		"webgl_geometry_terrain_fog_types",
```
Comment (EliasHasle) [inline examples/files.js]: The terrain generation is not considered an important aspect of this demo, but it demonstrates fog well. Should the example be renamed to emphasize fog more?

File: src/renderers/shaders/ShaderChunk/fog_fragment.glsl.js [hunk 6]
```diff
@@ -2,12 +2,23 @@ export default /* glsl */`
 #ifdef USE_FOG
 
 	#ifdef FOG_EXP2
-
-		float fogFactor = 1.0 - exp( - fogDensity * fogDensity * fogDepth * fogDepth );
+	
+		vec3 scaledFogPosition = fogDensity * fogPosition;
```
Comment (EliasHasle) [inline src/renderers/shaders/ShaderChunk/fog_fragment.glsl.js]: https://github.com/mrdoob/three.js/pull/17316#issuecomment-524620442

File: src/scenes/Fog.d.ts [hunk 7]
```diff
@@ -8,7 +8,7 @@ export interface IFog {
 }
 
 /**
- * This class contains the parameters that define linear fog, i.e., that grows linearly denser with the distance.
+ * This class contains the parameters that define standard fog, i.e., that grows smoothly denser from the near to the far distance.
```
Comment (EliasHasle) [inline src/scenes/Fog.d.ts]: Remember to update this too, when the doc page is updated.

File: test/unit/src/scenes/FogExp2.tests.js [hunk 8]
```diff
@@ -5,7 +5,7 @@
 
 import { FogExp2 } from '../../../../src/scenes/FogExp2';
 
-export default QUnit.module( 'FoxExp2', () => {
+export default QUnit.module( 'FogExp2', () => {
```
Comment (EliasHasle) [inline test/unit/src/scenes/FogExp2.tests.js]: This was a typo that I discovered when doing search&replace. It should be fixed whether or not the PR goes through.

**Review by EliasHasle — COMMENTED** [review]
The PR introduces FogExp in a straightforward, copy-paste, search-replace way.

The additional changes to documentation are needed. They may need a couple more critical eyes on them, though.

It felt natural to include the GLSL optimizations in the same PR. Comment https://github.com/mrdoob/three.js/pull/17316#issuecomment-524875838 by @WestLangley (and the following discussion) applies to this PR too, to the extent that it is a valid point:

> Be aware that this PR, as currently-written, is going to significantly change fog behavior for users who use linear fog with and an OrthographicCamera and a negative value for near plane.
> 
> I will not be surprised if there are complaints.

I replied that the docs do not seem to allow negative `near` plane, and that multiplying `sign(fogPosition.z)` into `fogDepth` may be a simple solution if negative `near` is to be supported. I suggest continuing the discussion here, since @supereggbert has expressed support for pursuing my PR as first choice.

File: docs/api/en/scenes/Fog.html (hunk 1 above)
Comment (EliasHasle) [reply docs/api/en/scenes/Fog.html]: https://www.google.com/search?q=plot+t+*+t+*+%283.0+-+2.0+*+t%29+from+0+to+1

File: test/unit/src/scenes/FogExp2.tests.js (hunk 8 above)
Comment (EliasHasle) [reply test/unit/src/scenes/FogExp2.tests.js]: Fixed.

File: editor/js/Sidebar.Scene.js (hunk 4 above)
Comment (EliasHasle) [reply editor/js/Sidebar.Scene.js]: Fixed.

File: editor/js/Sidebar.Scene.js (hunk 3 above)
Comment (EliasHasle) [reply editor/js/Sidebar.Scene.js]: It is no longer inconsistent when using the description "smooth ranged" in the docs.

File: docs/api/en/scenes/Fog.html (hunk 1 above)
Comment (EliasHasle) [reply docs/api/en/scenes/Fog.html]: Now described as "smooth ranged fog" with the latest commit. Any better name? Maybe this is good enough for now?

File: src/scenes/Fog.d.ts (hunk 7 above)
Comment (EliasHasle) [reply src/scenes/Fog.d.ts]: Up to date with "smooth ranged fog".

**Comment by EliasHasle** [comment]
@titansoftime Would you like to test this? Considering how #14786 ended.

File: docs/list.js (hunk 2 above)
Comment (gogoend) [reply docs/list.js]: Well,maybe you can just copy the En doc to the corresponding folder in Zh doc, then when the translators find it untranslated, they will directly translate on this file.

Comment (EliasHasle) [reply docs/list.js]: OK. I will trust that approach, then.

**Comment by EliasHasle** [comment]
@mrdoob Any chance of getting a review/merge soon? I try not to nag people too much, but FogExp was accepted by you as a good addition, so I went ahead with implementing it for common good. https://github.com/mrdoob/three.js/issues/17264#issuecomment-522270563 https://github.com/mrdoob/three.js/issues/17264#issuecomment-523213881

Are any of the additional changes problematic? I think I have justified them well, but I am sure more perspectives may improve them further. @Mugen87 @WestLangley

File: docs/api/en/scenes/Fog.html (hunk 1 above)
Comment (EliasHasle) [reply docs/api/en/scenes/Fog.html]: Decided with https://github.com/mrdoob/three.js/issues/17420#issuecomment-534588800 (changes needed here)

File: examples/files.js (hunk 5 above)
Comment (EliasHasle) [reply examples/files.js]: Not an important comment. Resolving to reduce noise.

**Comment by EliasHasle** [comment]
@titansoftime 
> I can test once merged.

Would you like to test the https://eliashasle.github.io/testing/fog_types/ demo on your mobile device before I start refactoring to the [new API}(https://github.com/mrdoob/three.js/issues/17420#issuecomment-534592079)?

I guess part of the refactoring will also be to make examples that explicitly set precision `mediump` and `lowp` for testing.

**Comment by titansoftime** [comment]
> Would you like to test the https://eliashasle.github.io/testing/fog_types/ demo on your mobile device before I start refactoring to the new API?

Looks good. Though without an orbit cam, it's kinda difficult to be sure. Previously I would rotate the camera to reveal "fail areas" in the fog.

**Comment by EliasHasle** [comment]
Thanks!

Did you get to test all three fog types and different param values? (dat.gui isn't the most mobile-friendly) And are you certain that you were running in `mediump`?

> without an orbit cam, it's kinda difficult to be sure.

I think I disabled it to avoid some event-handling conflicts.

**Comment by Mugen87** [comment]
Um, the diff from the PR looks strange. Can you please rebase it so it is based on the latest commit in `dev`?

**Comment by EliasHasle** [comment]
This gets worse and worse. Until a few minutes ago there was a monster diff, and now there is no diff at all, and the PR was "merged" and automatically closed. 😛 

I will just have to go on...

**Comment by EliasHasle** [comment]
#17592 is a new try... I will clean it up a bit before I mark it as ready for review.

**Pull request closed without merging**